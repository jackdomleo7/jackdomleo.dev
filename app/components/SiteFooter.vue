<template>
  <footer class="footer">
    <nuxt-link to="/" class="footer__logo">
      <Icon name="custom:j-icon" mode="svg" />
      <span class="sr-only">Home</span>
    </nuxt-link>
    <ul class="footer__quick-links">
      <li v-for="link in links" :key="link.key">
        <nuxt-link :to="link.value" :rel="link.value.startsWith('https://') ? 'noopener' : undefined">
          {{ link.key }}
        </nuxt-link>
      </li>
    </ul>
    <ul class="footer__social">
      <li v-for="socialLink in socialLinks" :key="socialLink.key">
        <nuxt-link :href="socialLink.value" target="_blank" rel="noopener noreferrer" data-cooltipz-dir="top" :aria-label="socialLink.key">
          <Icon class="footer__social-icon" :name="`custom:${socialLink.key.toLowerCase()}`" mode="svg" />
          <span class="sr-only">{{ socialLink.key }}</span>
        </nuxt-link>
      </li>
    </ul>
    <p class="footer__legal">
      <nuxt-link href="/privacy-policy" class="link">Privacy Policy</nuxt-link>
    </p>
    <div class="footer__legal">
      <p>Â©2018-{{ variables.CURRENT_YEAR }}</p>
      <p>All rights reserved.</p>
      <p>Jack Domleo</p></div>
  </footer>
</template>

<script lang="ts" setup>
import variables from '@/utilities/variables';

const links: { key: string; value: string }[] = [
  { key: 'Projects', value: '/projects' },
  { key: 'Work', value: '/work' },
  { key: 'Blog', value: '/blog' },
  { key: 'Links', value: '/links' }
]

const socialLinks: { key: string; value: string }[] = [
  { key: 'GitHub', value: 'https://github.com/jackdomleo7' },
  { key: 'CodePen', value: 'https://codepen.io/jackdomleo7' },
  { key: 'LinkedIn', value: 'https://linkedin.com/in/jackdomleo7' }
]
</script>

<style lang="scss" scoped>
.footer {
  max-width: 32rem;
  min-width: 100%;
  margin-top: 8rem;
  margin-inline: auto;
  padding: 2rem 1rem;
  display: flex;
  flex-direction: column;

  @media (min-width: $responsive-small-tablet) {
    padding-inline: 2rem;
    min-width: 32rem;
  }

  &__logo {
    width: 3.625rem;
    margin-inline: auto;
  }

  &__quick-links {
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 0.5rem;
    list-style-type: none;
    padding-left: 0;
    margin-block: 1.5rem;

    @media (min-width: $responsive-small-tablet) {
      gap: 1rem;
    }

    a {
      color: var(--color-primary);
      text-decoration: none;

      &:hover,
      &:active {
        text-decoration: underline;
        text-decoration-color: var(--color-accent);
        text-decoration-thickness: 2px;
      }
    }
  }

  &__social {
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 1rem;
    list-style-type: none;
    padding-left: 0;
    margin-top: 0;
    margin-bottom: 1.5rem;

    a {
      display: grid;
      place-items: center;
      border-radius: 50%;
      background-color: var(--color-bg);
      border: 1px solid var(--color-primary);
      padding: 0.375rem;
      color: var(--color-primary);
      width: 2.5rem;
      aspect-ratio: auto 1 / 1;
      outline-offset: 2px;
      
      @media (prefers-reduced-motion: no-preference) {
        transition: all var(--standard-animation-timing) ease;
      }
    }

    &-icon {
      height: 1.5rem;
      width: 1.5rem;
    }
  }

  &__legal {
    text-align: center;
    margin-block: 0;
    line-height: 1.3;

    @media (min-width: $responsive-small-tablet) {
      line-height: unset;
      display: flex;
      justify-content: center;
      flex-flow: wrap;
      gap: 0.25rem;
    }
  }
}
</style>