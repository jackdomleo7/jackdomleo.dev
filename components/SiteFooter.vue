<template>
  <footer class="footer">
    <nuxt-link to="/" class="footer__logo">
      <nuxt-icon name="j-icon" filled />
    </nuxt-link>
    <ul class="footer__quick-links">
      <li v-for="quickLink in data!.fields.quickLinks" :key="quickLink.fields.url">
        <nuxt-link :to="quickLink.fields.url" :rel="quickLink.fields.url.startsWith('https://') ? 'noopener' : undefined">
          {{ quickLink.fields.text }}
        </nuxt-link>
      </li>
    </ul>
    <ul class="footer__social">
      <li v-for="socialLink in data!.fields.socialLinks" :key="socialLink.fields.url">
        <nuxt-link :href="socialLink.fields.url" target="_blank" rel="noopener noreferrer">
          <nuxt-icon class="footer__social-icon" :name="socialLink.fields.icon" />
          <span class="sr-only">{{ socialLink.fields.text }}</span>
        </nuxt-link>
      </li>
    </ul>
    <div class="footer__legal" v-html="documentToHtmlString(data!.fields.legalText)"></div>
  </footer>
</template>

<script lang="ts" setup>
import { documentToHtmlString } from '@contentful/rich-text-html-renderer';
import { Contentful } from '@/enums/Contentful'
import type { ContentfulEntries } from '@/types/CMS/Entries'
import { formatCMSVariables } from '@/utilities/cmsVariables'

const { data } = await useAsyncData((ctx) => { return ctx!.$contentful.getEntry<ContentfulEntries.Footer>(Contentful.EntryIDs.FOOTER)})
data.value!.fields = formatCMSVariables(data.value!.fields)
</script>

<style lang="scss" scoped>
.footer {
  max-width: 32rem;
  min-width: 100%;
  margin-top: 8rem;
  margin-inline: auto;
  padding: 2rem 1rem;
  display: flex;
  flex-direction: column;

  @media (min-width: $responsive-small-tablet) {
    padding-inline: 2rem;
    min-width: 32rem;
  }

  &__logo {
    width: 3.625rem;
    margin-inline: auto;
  }

  &__quick-links {
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 0.5rem;
    list-style-type: none;
    padding-left: 0;
    margin-block: 1.5rem;

    @media (min-width: $responsive-small-tablet) {
      gap: 1rem;
    }

    a {
      color: var(--color-primary);
      text-decoration: none;

      &:hover,
      &:active {
        text-decoration: underline;
        text-decoration-color: var(--color-accent);
      }
    }
  }

  &__social {
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 1rem;
    list-style-type: none;
    padding-left: 0;
    margin-top: 0;
    margin-bottom: 1.5rem;

    a {
      display: grid;
      place-items: center;
      border-radius: 50%;
      overflow: hidden;
      background-color: var(--color-bg);
      border: 1px solid var(--color-primary);
      transition: all var(--standard-animation-timing) ease;
      padding: 0.375rem;
      color: var(--color-primary);
      width: 2.5rem;
      aspect-ratio: auto 1 / 1;
      outline-offset: 2px;
    }

    &-icon {
      height: 1.5rem;
      width: 1.5rem;
    }
  }

  &__legal {
    text-align: center;
    margin-block: 0;
    line-height: 1.3;

    @media (min-width: $responsive-small-tablet) {
      line-height: unset;
      display: flex;
      justify-content: center;
      flex-flow: wrap;
      gap: 0.25rem;
    }
  }
}
</style>