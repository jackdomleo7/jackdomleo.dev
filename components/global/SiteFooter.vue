<template>
  <footer class="footer">
    <form
      v-if="showNewsletterSubscribe"
      id="mc-embedded-subscribe-form"
      class="footer__subscribe"
      action="https://dev.us10.list-manage.com/subscribe/post?u=5db0c80170083cdaab0b17f84&amp;id=2f05546818"
      method="post"
      name="mc-embedded-subscribe-form"
      target="_blank"
      novalidate
    >
      <TextInput
        id="mce-EMAIL"
        v-model="newsletter.email"
        class="footer__subscribe-input"
        type="email"
        name="EMAIL"
        label="Newsletter"
        placeholder="Enter your email"
        help-text="No spam. Unsubscribe anytime."
        required
        hide-required-asterisk
        :error-msg="$v.newsletter.email.$dirty && $v.newsletter.email.$invalid ? 'Please provide a valid email' : undefined"
        @input="touch('email')"
        @blur="touch('email')"
      />
      <Btn class="footer__subscribe-btn" square type="submit" :disabled="$v.newsletter.$invalid" :aria-label="$v.newsletter.$invalid ? 'Please provide a valid email before subscribing' : undefined">
        Subscribe
      </Btn>
    </form>
    <nuxt-img src="~assets/images/j-icon.svg" alt="" class="footer__logo" />
    <ul class="footer__quick-links">
      <li v-for="quickLink in quickLinks" :key="quickLink.url">
        <a v-if="quickLink.url.startsWith('https://')" :href="quickLink.url" rel="noopener">
          {{ quickLink.text }}
        </a>
        <nuxt-link v-else :to="quickLink.url">
          {{ quickLink.text }}
        </nuxt-link>
      </li>
    </ul>
    <ul class="footer__social">
      <li v-for="socialLink in socialLinks" :key="socialLink.url">
        <a :href="socialLink.url" target="_blank" rel="noopener noreferrer">
          <svg-icon :name="socialLink.icon" />
          <span class="sr-only">{{ socialLink.text }}</span>
        </a>
      </li>
    </ul>
    <p class="footer__legal">
      &copy;{{ new Date().getFullYear() }}<br class="footer__legal-br" /> All rights reserved.<br class="footer__legal-br" /> Jack Domleo
    </p>
  </footer>
</template>

<script lang="ts">
import Vue from 'vue'
import { email, required } from 'vuelidate/lib/validators'
import { validationMixin } from 'vuelidate'
import TextInput from '../TextInput.vue'
import Btn from '../Btn.vue'

interface IQuickLink {
  text: string;
  url: string;
}

interface ISocialLink {
  text: string;
  url: string;
  icon: 'twitter' | 'linkedin' | 'github' | 'codepen'
}

export default Vue.extend({
  name: 'SiteFooter',
  components: { TextInput, Btn },
  mixins: [validationMixin],
  validations: {
    newsletter: {
      email: {
        email,
        required
      }
    }
  },
  data () {
    return {
      showNewsletterSubscribe: false,
      newsletter: {
        email: ''
      }
    }
  },
  computed: {
    quickLinks (): IQuickLink[] {
      return [
        {
          text: 'Home',
          url: '/'
        },
        {
          text: 'Projects',
          url: '/projects'
        },
        {
          text: 'Blog',
          url: '/blog'
        },
        {
          text: 'Book',
          url: '/products/level-up-your-career-today-developer-edition'
        }
      ]
    },
    socialLinks (): ISocialLink[] {
      return [
        {
          text: 'Twitter',
          url: 'https://twitter.com/jackdomleo7',
          icon: 'twitter'
        },
        {
          text: 'LinkedIn',
          url: 'https://linkedin.com/in/jackdomleo7/',
          icon: 'linkedin'
        },
        {
          text: 'GitHub',
          url: 'https://github.com/jackdomleo7',
          icon: 'github'
        },
        {
          text: 'CodePen',
          url: 'https://codepen.io/jackdomleo7',
          icon: 'codepen'
        }
      ]
    }
  },
  methods: {
    touch (formField: string): void {
      return this.$v.newsletter[formField]!.$touch()
    }
  }
})
</script>

<style lang="scss" scoped>
.footer {
  max-width: 32rem;
  min-width: 100%;
  margin-top: 8rem;
  margin-inline: auto;
  padding: 2rem 1rem;
  display: flex;
  flex-direction: column;

  @media (min-width: $responsive-small-tablet) {
    padding-inline: 2rem;
    min-width: 32rem;
  }

  &__subscribe {
    display: flex;
    flex-direction: column;
    gap: 0.5rem;
    margin-bottom: 4rem;

    @media (min-width: $responsive-small-tablet) {
      flex-direction: row;
      margin-bottom: 7.5rem;
    }

    &-input {
      flex: 1;
    }

    &-btn {
      min-width: unset;

      @media (min-width: $responsive-small-tablet) {
        position: relative;
        top: 2.125rem;
      }
    }
  }

  &__logo {
    width: 3.625rem;
    margin-inline: auto;
  }

  &__quick-links {
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 0.5rem;
    list-style-type: none;
    padding-left: 0;
    margin-block: 1.5rem;

    @media (min-width: $responsive-small-tablet) {
      gap: 1rem;
    }

    a {
      color: var(--colour-text-primary);
    }
  }

  &__social {
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 1rem;
    list-style-type: none;
    padding-left: 0;
    margin-top: 0;
    margin-bottom: 1.5rem;

    a {
      display: grid;
      place-items: center;
      border-radius: 50%;
      overflow: hidden;
      background-color: var(--colour-background);
      border: 1px solid var(--colour-text-primary);
      transition: all var(--standard-animation-timing) ease;
      padding: 0.375rem;
      color: var(--colour-text-primary);
      width: 2.5rem;
      aspect-ratio: auto 1 / 1;

      &:hover,
      &:focus {
        box-shadow: var(--shadow-standard);
      }
    }
  }

  &__legal {
    text-align: center;
    margin-block: 0;
    line-height: 1.3;

    @media (min-width: $responsive-small-tablet) {
      line-height: unset;
    }

    &-br {
      @media (min-width: $responsive-small-tablet) {
        display: none;
      }
    }
  }
}
</style>
