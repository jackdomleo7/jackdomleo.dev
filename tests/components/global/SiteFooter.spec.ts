import MockDate from 'mockdate'
import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'
import SiteFooter from '@/components/global/SiteFooter.vue'

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because components are loaded in isolation
  }
})

describe('SiteFooter', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(SiteFooter, {
      stubs: ['nuxt-link', 'svg-icon', 'nuxt-img']
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(SiteFooter, {
      stubs: ['nuxt-link', 'svg-icon', 'nuxt-img']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(SiteFooter, {
      stubs: ['nuxt-link', 'svg-icon', 'nuxt-img']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(SiteFooter, {
      stubs: ['nuxt-link', 'svg-icon', 'nuxt-img']
    })

    await expect(axe(wrapper.element)).resolves.toHaveNoViolations()
  })

  /**
   * Interactivity
   */

  // Skipped while newsletter signup is hidden
  it.skip('enables the "Subscibe" button when a valid email is entered into the Newsletter input', async () => {
    const wrapper = mount(SiteFooter, {
      stubs: ['nuxt-link', 'svg-icon', 'nuxt-img']
    })

    expect(wrapper.find('button.footer__subscribe-btn').attributes('disabled')).toBe('disabled')

    await wrapper.find('input#mce-EMAIL').setValue('test@example.com')

    expect(wrapper.find('button.footer__subscribe-btn').attributes('disabled')).toBeUndefined()

    await wrapper.find('input#mce-EMAIL').setValue('test@example')

    expect(wrapper.find('button.footer__subscribe-btn').attributes('disabled')).toBe('disabled')

    await wrapper.find('input#mce-EMAIL').setValue('')

    expect(wrapper.find('button.footer__subscribe-btn').attributes('disabled')).toBe('disabled')
  })

  /**
   * Misc
   */

  it('shows the correct year in the legal text', () => {
    MockDate.set('2022-01-01')

    const year = 2022

    const wrapper = mount(SiteFooter, {
      stubs: ['nuxt-link', 'svg-icon', 'nuxt-img']
    })

    expect(wrapper.find('p.footer__legal').text()).toContain(String(year))
  })
})
