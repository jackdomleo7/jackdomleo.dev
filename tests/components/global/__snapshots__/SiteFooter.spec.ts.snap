// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteFooter matches snapshot when mounted 1`] = `
<footer
  class="footer"
>
  <!---->
   
  <nuxt-img-stub
    alt=""
    class="footer__logo"
    src="~assets/images/j-icon.svg"
  />
   
  <ul
    class="footer__quick-links"
  >
    <li>
      <nuxt-link-stub
        to="/"
      >
        
        Home
      
      </nuxt-link-stub>
    </li>
    <li>
      <nuxt-link-stub
        to="/projects"
      >
        
        Projects
      
      </nuxt-link-stub>
    </li>
    <li>
      <nuxt-link-stub
        to="/blog"
      >
        
        Blog
      
      </nuxt-link-stub>
    </li>
    <li>
      <nuxt-link-stub
        to="/products/level-up-your-career-today-developer-edition"
      >
        
        Book
      
      </nuxt-link-stub>
    </li>
  </ul>
   
  <ul
    class="footer__social"
  >
    <li>
      <a
        href="https://twitter.com/jackdomleo7"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="twitter"
        />
         
        <span
          class="sr-only"
        >
          Twitter
        </span>
      </a>
    </li>
    <li>
      <a
        href="https://linkedin.com/in/jackdomleo7/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="linkedin"
        />
         
        <span
          class="sr-only"
        >
          LinkedIn
        </span>
      </a>
    </li>
    <li>
      <a
        href="https://github.com/jackdomleo7"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="github"
        />
         
        <span
          class="sr-only"
        >
          GitHub
        </span>
      </a>
    </li>
    <li>
      <a
        href="https://codepen.io/jackdomleo7"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="codepen"
        />
         
        <span
          class="sr-only"
        >
          CodePen
        </span>
      </a>
    </li>
  </ul>
   
  <p
    class="footer__legal"
  >
    
    ©2022
    <br
      class="footer__legal-br"
    />
     All rights reserved.
    <br
      class="footer__legal-br"
    />
     Jack Domleo
  
  </p>
</footer>
`;

exports[`SiteFooter matches snapshot when shallow mounted 1`] = `
<footer
  class="footer"
>
  <!---->
   
  <nuxt-img-stub
    alt=""
    class="footer__logo"
    src="~assets/images/j-icon.svg"
  />
   
  <ul
    class="footer__quick-links"
  >
    <li>
      <nuxt-link-stub
        to="/"
      >
        
        Home
      
      </nuxt-link-stub>
    </li>
    <li>
      <nuxt-link-stub
        to="/projects"
      >
        
        Projects
      
      </nuxt-link-stub>
    </li>
    <li>
      <nuxt-link-stub
        to="/blog"
      >
        
        Blog
      
      </nuxt-link-stub>
    </li>
    <li>
      <nuxt-link-stub
        to="/products/level-up-your-career-today-developer-edition"
      >
        
        Book
      
      </nuxt-link-stub>
    </li>
  </ul>
   
  <ul
    class="footer__social"
  >
    <li>
      <a
        href="https://twitter.com/jackdomleo7"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="twitter"
        />
         
        <span
          class="sr-only"
        >
          Twitter
        </span>
      </a>
    </li>
    <li>
      <a
        href="https://linkedin.com/in/jackdomleo7/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="linkedin"
        />
         
        <span
          class="sr-only"
        >
          LinkedIn
        </span>
      </a>
    </li>
    <li>
      <a
        href="https://github.com/jackdomleo7"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="github"
        />
         
        <span
          class="sr-only"
        >
          GitHub
        </span>
      </a>
    </li>
    <li>
      <a
        href="https://codepen.io/jackdomleo7"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg-icon-stub
          name="codepen"
        />
         
        <span
          class="sr-only"
        >
          CodePen
        </span>
      </a>
    </li>
  </ul>
   
  <p
    class="footer__legal"
  >
    
    ©2022
    <br
      class="footer__legal-br"
    />
     All rights reserved.
    <br
      class="footer__legal-br"
    />
     Jack Domleo
  
  </p>
</footer>
`;
